actpool/actqueue.go

```diff
@@ -27,24 +27,26 @@ type nonceWithTTL struct {
 	deadline time.Time
 }
 
-type noncePriorityQueue []nonceWithTTL
+type noncePriorityQueue []*nonceWithTTL
 
 func (h noncePriorityQueue) Len() int           { return len(h) }
 func (h noncePriorityQueue) Less(i, j int) bool { return h[i].nonce < h[j].nonce }
 func (h noncePriorityQueue) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
 
 func (h *noncePriorityQueue) Push(x interface{}) {
-	in, ok := x.(nonceWithTTL)
-	if !ok {
-		return
+	if in, ok := x.(*nonceWithTTL); ok {
+		*h = append(*h, in)
 	}
-	*h = append(*h, in)
 }
 
 func (h *noncePriorityQueue) Pop() interface{} {
 	old := *h
 	n := len(old)
+	if n == 0 {
+		return nil
+	}
 	x := old[n-1]
+	old[n-1] = nil // avoid memory leak
```

Comment:
how could memory leak happen? why the memory gc cannot detect it?

Reply:
>As mentioned earlier, re-slicing a slice doesn’t make a copy of the underlying array. The full array will be kept in memory until it is no longer referenced. Occasionally this can cause the program to hold all the data in memory when only a small piece of it is needed.

https://go.dev/blog/slices-intro

Reply:
I must be dumb, could you point me to the exact place or line where explain the details?

Reply:
Got it.
How will this change (struct to pointer) help code efficiency?

Reply:
It should. The pointer implementation is referred from official pq (https://github.com/golang/go/blob/master/src/container/heap/example_pq_test.go).

Reply:
not an explanation.